src/bin/wasmtime.rs

```diff
@@ -1,457 +1,81 @@
-                exit(1);
-            }
-        }
-    }
-
-    let errors = cache_init(
-        !args.flag_disable_cache,
-        args.flag_cache_config.as_ref(),
-        log_config,
-    );
-
-    if !errors.is_empty() {
-        eprintln!("Cache initialization failed. Errors:");
-        for e in errors {
-            eprintln!("-> {}", e);
-        }
-        exit(1);
-    }
-
-    let mut flag_builder = settings::builder();
-    let mut features: Features = Default::default();
-
-    // There are two possible traps for division, and this way
-    // we get the proper one if code traps.
-    flag_builder.enable("avoid_div_traps")?;
-
-    // Enable/disable producing of debug info.
-    let debug_info = args.flag_g;
-
-    // Enable verifier passes in debug mode.
-    if cfg!(debug_assertions) {
-        flag_builder.enable("enable_verifier")?;
+    let mut app = App::new("wasmtime")
+        .version(env!("CARGO_PKG_VERSION"))
+        .about("Wasmtime WebAssembly Runtime")
+        .after_help("If a subcommand is not provided, the `run` subcommand will be used.")
+        .setting(AppSettings::VersionlessSubcommands)
+        .setting(AppSettings::SubcommandRequiredElseHelp)
+        .setting(AppSettings::ColoredHelp);
+
+    for (_, factory, _) in SUBCOMMANDS {
+        app = app.subcommand(factory());
     }
 
-    // Enable SIMD if requested
-    if args.flag_enable_simd {
-        flag_builder.enable("enable_simd")?;
-        features.simd = true;
-    }
-
-    // Enable optimization if requested.
-    if args.flag_optimize {
-        flag_builder.set("opt_level", "speed")?;
-    }
-
-    // Decide how to compile.
-    let strategy = pick_compilation_strategy(args.flag_cranelift, args.flag_lightbeam);
-
-    let mut config = Config::new();
-    config
-        .features(features)
-        .flags(settings::Flags::new(flag_builder))
-        .debug_info(debug_info)
-        .strategy(strategy);
-    let engine = HostRef::new(Engine::new(&config));
-    let store = HostRef::new(Store::new(&engine));
-
-    let mut module_registry = HashMap::new();
-
-    // Make wasi available by default.
-    let preopen_dirs = compute_preopen_dirs(&args.flag_dir, &args.flag_mapdir);
-    let argv = compute_argv(&args.arg_file, &args.arg_arg);
-    let environ = compute_environ(&args.flag_env);
-
-    let wasi_unstable = HostRef::new(if args.flag_wasi_c {
-        #[cfg(feature = "wasi-c")]
-        {
-            let global_exports = store.borrow().global_exports().clone();
-            let handle = instantiate_wasi_c(global_exports, &preopen_dirs, &argv, &environ)?;
-            Instance::from_handle(&store, handle)
-        }
-        #[cfg(not(feature = "wasi-c"))]
-        {
-            bail!("wasi-c feature not enabled at build time")
-        }
-    } else {
-        create_wasi_instance_snapshot_0(&store, &preopen_dirs, &argv, &environ)?
+    let matches = app.get_matches_safe().unwrap_or_else(|e| match e.kind {
```

Comment:
It might be easier to do this auto-run dance by collecting the CLI args in a `Vec` and then manually inserting `run` if necessary.

```rust
let mut args: Vec<_> = std::env::args().collect();
match args.get(1) {
    None => app.show_help_text_and_exit(),
    Some(a) if is_wasm_file(a) => args.insert(1, "run".to_string()),
    _ => {}
}
let args = app.get_matches_from_iter(&args);
```

Something like that.

Reply:
The problem with this approach is that it mandates that the first argument must be the file to run and it would need to be recognizable as a .wasm file (or alternatively not match a known subcommand).

For example, `wasmtime -g foo.wasm` would not work.  Checking for known subcommands might work, but then it prevents us from having any universal top-level options in the future (e.g. `wasmtime --top-level-option run foo.wasm`).

I figured a robust, although more complex, solution would be to parse the command line once and then parse it again upon failure as if it were a run command.

I'm open to a different approach, though.

Reply:
Ah, that makes a lot of sense, and I didn't think of `--top-level-option` kinds of things. Given that, I think this is the way to go!

Reply:
On the plus side, it does look cleaner with `structopt`.